RavenDB-11373 Recording and replaying a sequence of MergedTransaction…

ayende opened the pull request:
…Commands in relation to the transaction in which they were written to the disk including end points & client commands and operations + test

haludi committed:
RavenDB-11373 Recording and replaying a sequence of MergedTransactionCommands in relation to the transaction in which they were written to the disk including end points & client commands and operations + test

ayende commented on src/Raven.Client/Json/BlittableJsonWriter.cs:
```diff
@@ -550,6 +550,25 @@ public override void WriteValue(Uri value)
+            switch (value)
+            {
+                case BlittableJsonReaderObject readerObject:
+                    if (false == readerObject.HasParent)
```
You did too much C :-)

ayende commented on src/Raven.Client/Json/Converters/BlittableJsonConverter.cs:
```diff
@@ -0,0 +1,53 @@
+                        ? blittableValue
+                        : blittableValue.Clone(blittableReader.Context);
+            }
+            throw new SerializationException($"Try to read {nameof(BlittableJsonReaderObject)} from non {nameof(BlittableJsonReaderObject)} value");
```
What _is_ the value that you go? We need to provide as many details as possible to make it easy for the user to understand.

haludi replied:
done

ayende commented on src/Raven.Client/Json/Converters/BlittableJsonConverter.cs:
```diff
@@ -0,0 +1,53 @@
+        {
+            if (!(reader is BlittableJsonReader blittableReader))
+            {
+                throw new SerializationException($"{nameof(BlittableJsonReader)} must to be used for convert to {nameof(BlittableJsonReaderObject)}");
```
Please provide good error here on what value _was_ passed here.

haludi replied:
done

ayende commented on src/Raven.Client/Json/Converters/LazyStringValueJsonConverter.cs:
```diff
@@ -0,0 +1,42 @@
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            if (!(reader is BlittableJsonReader blittableReader))
```
Better error on what was passed here.

haludi replied:
done

ayende commented on src/Raven.Client/Json/Converters/LazyStringValueJsonConverter.cs:
```diff
@@ -0,0 +1,42 @@
+                return null;
+            }
+
+            //Todo It will be better to change the reader to set the value as LazyStringValue 
```
What if the value is _already_ `LazyBlittableString` ?

haludi replied:
https://github.com/haludi/ravendb/blob/f13bbd5c65c488aa7c227f6de329e735570a3d44/src/Raven.Client/Json/BlittableJsonReader.cs#L205

The reader convert all the serialized string types to string before it set the value

ayende commented on src/Raven.Client/Json/Converters/StreamConverter.cs:
```diff
@@ -0,0 +1,59 @@
+        {
+            if (!(writer is BlittableJsonWriter blittableJsonWriter))
+            {
+                throw new SerializationException($"{nameof(BlittableJsonWriter)} must to be used for serialize {nameof(Stream)}");
```
Better error specify what was the value we did get here

haludi replied:
done

ayende commented on src/Raven.Client/Json/Converters/StreamConverter.cs:
```diff
@@ -0,0 +1,59 @@
+
+            if (!(value is Stream stream))
+            {
+                throw new SerializationException($"Try to write non {nameof(Stream)} value with {nameof(StreamConverter)}");
```
What value did we get here?

haludi replied:
done

ayende commented on src/Raven.Client/Json/Converters/StreamConverter.cs:
```diff
@@ -0,0 +1,59 @@
+                throw new SerializationException($"Try to write non {nameof(Stream)} value with {nameof(StreamConverter)}");
+            }
+
+            //Todo To consider if Seek need to be here or should it come ready
```
You do need it here, and you need to check that you `CanSeek` and error appropriately.
Because you consume the full data, the serialization mutates the state of the object. We want to return it to `0` after the call.

haludi replied:
done

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -1767,8 +1767,7 @@ public static LazyStringValue TableValueToId(JsonOperationContext context, int i
         {
             // See format of the lazy string ID in the GetLowerIdSliceAndStorageKey method
```
Comment needs to move to `GetLazyStringValue` method

haludi replied:
done

ayende commented on src/Raven.Server/Documents/Expiration/ExpiredDocumentsCleaner.cs:
```diff
@@ -123,10 +124,30 @@ public DeleteExpiredDocumentsCommand(Dictionary<Slice, List<(Slice LowerId, Lazy
+    {
+        public ExpiredDocumentsCleaner.DeleteExpiredDocumentsCommand ToCommand(DocumentsOperationContext context, DocumentDatabase database)
+        {
+            Dictionary<Slice, List<(Slice LowerId, LazyStringValue Id)>> expired = Expired.ToDictionary(kv => kv.Key, kv => kv.Value);
```
Why not clone it?

haludi replied:
https://www.newtonsoft.com/json/help/html/SerializationGuide.htm#Dictionarys

At the time I did not look into it in depth, but using the dictionary rose a problem
Now I've found that there is special regard to the key of the dictionary
that cause it not going through the SliceJsonConverter

ayende replied:
Use `new Dictionary<...>(Expires)`

`ToDictionary` is invoking two delegates per instance and can throw if there are duplicates (which isn't an issue when reading from a dictionary, but in that case, why bother with `ToDictionary` ?

haludi replied:
done

ayende commented on src/Raven.Server/Documents/Handlers/BatchHandler.cs:
```diff
@@ -948,4 +966,47 @@ private class WaitForIndexItem
+        {
+            if (ParsedCommands == null || ParsedCommands.Any() == false)
+            {
+                //Todo To check if empty array can append & if it is make sense to check it here 
```
You don't need the check, we should never get an empty array here.
If we do get it, however, just handle it normally.

haludi replied:
done

ayende commented on src/Raven.Server/Documents/Handlers/BulkInsertHandler.cs:
```diff
@@ -206,6 +210,38 @@ public override int Execute(DocumentsOperationContext context)
+        {
+            if (Commands == null || Commands.Any() == false) 
+            {
+                //Todo To check if empty array can append & if it is make sense to check it here 
```
Don't need to check here.

haludi replied:
done

ayende commented on src/Raven.Server/Documents/Handlers/TransactionsRecordingHandler.cs:
```diff
@@ -0,0 +1,173 @@
+                    foreach (var replayProgress in Database.TxMerger.Replay(replayStream))
+                    {
+                        commandsProgress = replayProgress.CommandsProgress;
+                        //TODO Maybe should be relative to time or size
```
We typically use the number of operations to push this.
This give the user predictable behavior in most cases, this is fine

haludi replied:
So how to decide how many operations should it be?

ayende replied:
1024

haludi replied:
done

ayende commented on src/Raven.Server/Documents/Handlers/TransactionsRecordingHandler.cs:
```diff
@@ -0,0 +1,173 @@
+            }
+            catch (Exception e)
+            {
+                HttpContext.Response.Headers["Connection"] = "close";
```
Add a comment on why this is required.

haludi replied:
done

ayende commented on src/Raven.Server/Documents/Patch/PatchDocumentCommand.cs:
```diff
@@ -25,8 +29,9 @@ public class PatchDocumentCommand : TransactionOperationsMerger.MergedTransactio
-        private readonly BlittableJsonReaderObject _patchIfMissingArgs;
-        private readonly BlittableJsonReaderObject _patchArgs;
+
+        private readonly (PatchRequest run, BlittableJsonReaderObject args) _patchIfMissing;
```
We use `PascalCase` for tuple values.

haludi replied:
You talk about the tuple properties, right?

ayende replied:
`(PatchRequest Run, BlittableJsonReaderObject Args)`
Yes

haludi replied:
done

ayende commented on src/Raven.Server/Documents/Queries/AbstractQueryRunner.cs:
```diff
@@ -169,7 +169,8 @@ private static IndexQueryServerSide ConvertToOperationQuery(IndexQueryServerSide
         }
 
-        private class BulkOperationCommand<T> : TransactionOperationsMerger.MergedTransactionCommand where T : TransactionOperationsMerger.MergedTransactionCommand
+        //Todo To check if it make sense to changed the accessibility of this type for adding it to the test list and if not to find alternative
```
This is fine

haludi replied:
done

ayende commented on src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:
```diff
@@ -1317,6 +1348,167 @@ public void Dispose()
+
+            public override TransactionOperationsMerger.IReplayableCommandDto<TransactionOperationsMerger.MergedTransactionCommand> ToDto(JsonOperationContext context)
+            {
+                var buffer = new byte[_totalSize];
```
This is a short lived buffer that may be large. Let's use `ArrayPool` here

haludi replied:
Issue created
[RavenDB-11852](http://issues.hibernatingrhinos.com/issue/RavenDB-11852)

ppekrol commented on src/Raven.Client/Documents/Operations/TransactionsRecording/ReplayTrxOperationResult.cs:
```diff
@@ -0,0 +1,23 @@
+
+namespace Raven.Client.Documents.Operations.TransactionsRecording
+{
+    public class ReplayTrxOperationResult : IOperationResult
```
We do not use 'Trx'. Change to 'Tx'

haludi replied:
done

ppekrol commented on src/Raven.Client/Documents/Operations/TransactionsRecording/ReplayTrxProgress.cs:
```diff
@@ -0,0 +1,21 @@
+
+namespace Raven.Client.Documents.Operations.TransactionsRecording
+{
+    public class ReplayTrxProgress : IOperationProgress
```
We do not use 'Trx'. Change to 'Tx'

haludi replied:
done

ayende commented on src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:
```diff
@@ -1317,6 +1348,167 @@ public void Dispose()
+            {
+                var buffer = new byte[_totalSize];
+                Marshal.Copy((IntPtr)_buffer, buffer, 0, _totalSize);
+                var strBuffer = Convert.ToBase64String(buffer);
```
This may be big, we need to investigate if we can do this in unmanaged memory more easily. 
`Sparrow.Utils.Base64` give you the option to go from `byte*` of buffer to `byte*` of base 64 without allocating managed memory.

haludi replied:
Issue created
[RavenDB-11852](http://issues.hibernatingrhinos.com/issue/RavenDB-11852)

ayende commented on src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:
```diff
@@ -1317,6 +1348,167 @@ public void Dispose()
+
+            unsafe
+            {
+                var buffer = Convert.FromBase64String(Buffer);
```
Same here, let's avoid these big allocations.

haludi replied:
Issue created
[RavenDB-11852](http://issues.hibernatingrhinos.com/issue/RavenDB-11852)

ayende commented on src/Raven.Server/Documents/TransactionOperationsMerger.cs:
```diff
@@ -65,9 +91,35 @@ public void Start()
+
+        public interface IReplayableCommandDto<out T> where T : MergedTransactionCommand
+        {
+            T ToCommand(DocumentsOperationContext context, DocumentDatabase database); // receive some parameters here?
```
Can remove the comment, I think

haludi replied:
done

ayende commented on src/Raven.Server/Documents/TransactionOperationsMerger.cs:
```diff
@@ -110,6 +162,178 @@ private static void ThrowTxMergerWasDisposed()
+
+                public void Init(string filePath)
+                {
+                    var recordingFileStream = new FileStream(filePath, FileMode.Create);
```
If we provide a bad file path (`f:\foo`, no permissions, file already exists, etc). What will happen?

haludi replied:
I choose to move the ownership of the record stream to the TxMerger and create the file while the start record request so the error will be arrived from there.
As I implemented it now the record will overwrite the existing file.
I can think about three ways to handle it:
1. To overwrite
2. To return error
3. To add an extension (maybe current time) that differentiate it from the existing

ayende commented on src/Raven.Server/Documents/TransactionOperationsMerger.cs:
```diff
@@ -848,5 +1125,283 @@ public void Dispose()
+                            }
+                            catch (Exception)
+                            {
+                                //TODO To accept exceptions that was thrown while recording
```
Reminder on this

haludi replied:
[RavenDB-11751](http://issues.hibernatingrhinos.com/issue/RavenDB-11751)

ayende commented on src/Raven.Server/Documents/TransactionOperationsMerger.cs:
```diff
@@ -848,5 +1125,283 @@ public void Dispose()
+
+        private IReplayableCommandDto<MergedTransactionCommand> DeserializeCommandDto(string type, JsonSerializer jsonSerializer, BlittableJsonReader reader, PeepingTomStream peepingTomStream)
+        {
+            //Todo Maybe should be simplified by convention & reflection
```
This is fine

haludi replied:
done

ayende commented on src/Raven.Server/Documents/TransactionOperationsMerger.cs:
```diff
@@ -848,5 +1125,283 @@ public void Dispose()
+                }
+            }
+
+            public override string Message => base.Message + ";\rContext:\r" + Context;
```
Use `\r\n` or `Environment.NewLine` here.

haludi replied:
done

ayende commented on src/Raven.Server/Json/Converters/SliceJsonConverter.cs:
```diff
@@ -0,0 +1,72 @@
+        {
+            if (!(writer is BlittableJsonWriter blittableJsonWriter))
+            {
+                throw new SerializationException($"{nameof(BlittableJsonWriter)} must to be used for write {nameof(LazyStringValue)}");
```
Better error here, what did we get?

haludi replied:
done

ayende commented on src/Raven.Server/Json/Converters/SliceJsonConverter.cs:
```diff
@@ -0,0 +1,72 @@
+
+            if (!(value is Slice slice))
+            {
+                throw new SerializationException($"Try to write non {nameof(Slice)} value with {nameof(SliceJsonConverter)}");
```
What did we get?

haludi replied:
done

ayende commented on src/Raven.Server/Json/Converters/SliceJsonConverter.cs:
```diff
@@ -0,0 +1,72 @@
+                throw new SerializationException($"Try to write non {nameof(Slice)} value with {nameof(SliceJsonConverter)}");
+            }
+
+            var buffer = new byte[slice.Size];
```
Base we use `Base64` and avoid the managed allocations?

haludi replied:
Issue created
[RavenDB-11852](http://issues.hibernatingrhinos.com/issue/RavenDB-11852)

ayende commented on src/Raven.Server/Json/Converters/SliceJsonConverter.cs:
```diff
@@ -0,0 +1,72 @@
+        {
+            if (!(reader is BlittableJsonReader blittableReader))
+            {
+                throw new SerializationException($"{nameof(BlittableJsonReader)} must to be used for convert to {nameof(BlittableJsonReaderObject)}");
```
Better error

haludi replied:
done

ayende commented on src/Raven.Server/Json/Converters/SliceJsonConverter.cs:
```diff
@@ -0,0 +1,72 @@
+            }
+
+            //TODO To create slice by JsonOperationContext
+            if (!(blittableReader.Context is DocumentsOperationContext context))
```
Why do you need `DocumentsOperationContext` and not just `JsonOperationContext`?

haludi replied:
[`Slice.From(context.Allocator, buffer, ByteStringType.Immutable, out var slice);`](https://github.com/ravendb/ravendb/pull/7397/files/1ab8eb87a068abb6e93c176b114e45468a6c1b1b#diff-d9907d27cc6f398f4ccad4f9d7d9823bR62)

As I understand it, only the TransactionOperationContext hold the Allocator that I need to create Slice. 
I can create it by myself but in this case, I will need to dispose it.
For that, I need to bring the allocator from the outside, the reader in this case,
and release the memory after the replay is done (or after the command execute).
It seems to me very messy to hold two objects (context & allocator) than handle memory in the reader.
The other option I can think about is to hold an allocator in the JsonOperationContext but this is a huge change to the system. Maybe lazy allocator?
Or there is something I didn't think about...

ayende replied:
No, that is fine.

ayende commented on src/Sparrow/Json/BlittableJsonReaderObject.cs:
```diff
@@ -782,6 +782,13 @@ private static void InvalidAttemptToCopyNestedObject()
 
+        //There is an issue when write blittable & at the same time clone another blittable on the same context
+        //Todo To check if should remove this method after issue is fixed
+        public BlittableJsonReaderObject CloneOnTheSameContext()
```
What is the status of this?

ayende commented on src/Sparrow/Json/JsonOperationContext.cs:
```diff
@@ -526,6 +526,13 @@ private unsafe LazyStringValue GetLazyString(StringSegment field, bool longLived
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public unsafe LazyStringValue GetLazyStringValue(byte* ptr)
+        {
+            var size = BlittableJsonReaderBase.ReadVariableSizeInt(ptr, 0, out var offset);
```
Bring the comment on the structure here.

haludi replied:
done

ppekrol commented on src/Raven.Server/Documents/Handlers/TransactionsRecordingHandler.cs:
```diff
@@ -0,0 +1,173 @@
+        [RavenAction("/databases/*/transactions/replay", "POST", AuthorizationStatus.ValidUser)]
+        public async Task ReplayRecording()
+        {
+            var operationId = GetLongQueryString("operationId");
```
The operation should not be passed from the outside (unless used via Studio).

This endpoint should return the OperationId only (check other similar methods e.g. in SmugglerHandler)

The ReplayOperation in client should return OperationIdResult

haludi replied:
As I understand it, the request is executed while the file stream is uploaded.
So, if the OperationId will be passed at the end of the request the client will not be able to monitor the operation that can take a lot of time.

ppekrol commented on src/Raven.Server/Documents/Handlers/TransactionsRecordingHandler.cs:
```diff
@@ -0,0 +1,173 @@
+            }
+        }
+
+        private async Task<IOperationResult> DoReplay(
```
This method does not have to be async?

haludi replied:
done

ppekrol commented on src/Raven.Server/Documents/Handlers/TransactionsRecordingHandler.cs:
```diff
@@ -0,0 +1,173 @@
+                    ThrowRequiredPropertyNameInRequest(nameof(parameters.File));
+                }
+
+                var command = new TransactionsRecordingCommand(
```
isn't this an overkill? can we do `Database.TxMerger.Start(parameters.File)` ?

ayende replied:
We have to do it this way to avoid starting recording in the middle of a transaction.
This way, we always start from a brand new tx.

ppekrol replied:
Makes sense.

ppekrol commented on src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:
```diff
@@ -1317,6 +1348,167 @@ public void Dispose()
+            var replicationItems = ReplicatedItemDtos.Select(d => d.ToItem(context)).ToList();
+
+            var replicatedAttachmentStreams = ReplicatedAttachmentStreams.Select(i => CreateReplicationAttachmentStream(context, i)).ToArray();
+            var replicationHandler = new IncomingReplicationHandler(database, replicationItems, connectionInfo, database.ReplicationLoader, replicatedAttachmentStreams);
```
Can we avoid doing that?

haludi replied:
Avoiding what? Linq?

maximburyak replied:
@haludi I think @ppekrol was referencing the creationg of IncomingReplicationHandler, which is pretty heavy object, that does a lot of stuff under the hood. I think that at this stage it will be a bit hard to get rid of it, because of the none trivial code in MergedDocumentReplicationCommand.Execute. But yes, it looks like a big TODO

haludi replied:
An issue created
[RavenDB-11857](http://issues.hibernatingrhinos.com/issue/RavenDB-11857)

ppekrol commented on src/Raven.Server/Documents/Replication/ResolveConflictOnReplicationConfigurationChange.cs:
```diff
@@ -161,6 +163,21 @@ public override int Execute(DocumentsOperationContext context)
+                // so EscapePositions set to empty to avoid reference to escape bytes (after string bytes) while serializing
+                foreach (var conflict in _resolvedConflicts)
+                {
+                    conflict.ResolvedConflict.LowerId.EscapePositions = Array.Empty<int>();
```
Don't we need to make a copy of them then?

haludi replied:
In memory LazyStringValue doesn't have escape characters.
Maybe I will change the comment?

> The LowerId created as in memory LazyStringValue that <b> doesn't have escape characters </b>
so EscapePositions set to empty to avoid reference to escape bytes (after string bytes) while serializing

ayende replied:
change the comment, yes

haludi replied:
done

ppekrol commented on src/Raven.Server/Documents/TransactionOperationsMerger.cs:
```diff
@@ -848,5 +1125,283 @@ public void Dispose()
+            public StartRecordingDetails()
+            : base(DetailsType)
+            {
+                Version = ServerVersion.FullVersion;
```
We typically use RavenVersionAttribute.Instance.Build only

haludi replied:
done

ppekrol commented on src/Raven.Client/Documents/Conventions/DocumentConventions.cs:
```diff
@@ -705,6 +705,9 @@ public JsonSerializer CreateSerializer()
             jsonSerializer.Converters.Add(JsonIMetadataDictionaryConverter.Instance);
+            jsonSerializer.Converters.Add(BlittableJsonConverter.Instance);
+            jsonSerializer.Converters.Add(LazyStringValueJsonConverter.Instance);
+            jsonSerializer.Converters.Add(StreamConverter.Instance);
```
Not sure if we want to have these as default in client? Those are used only on server-side?

ayende replied:
Yes, it is better to avoid them there by default.

haludi replied:
done

ppekrol commented on src/Raven.Client/Documents/Operations/TransactionsRecording/ReplayTransactionsRecordingOperation.cs:
```diff
@@ -0,0 +1,64 @@
+
+namespace Raven.Client.Documents.Operations.TransactionsRecording
+{
+    public class ReplayTransactionsRecordingOperation : IMaintenanceOperation<ReplayTrxOperationResult>
```
As written below. This needs to return the OperationIdResult

haludi replied:
https://github.com/ravendb/ravendb/pull/7397#discussion_r212181012

ppekrol commented on src/Raven.Server/Documents/TransactionOperationsMerger.cs:
```diff
@@ -848,5 +1125,283 @@ public void Dispose()
+        }
+
+        public IEnumerable<ReplayProgress> Replay(Stream replayStream)
+        {
```
Can we move the implementation to some dedicated helper class?

haludi replied:
done

haludi committed:
Compilation errors after cherry pick solved

haludi committed:
RavenDB-11373
* new converters removed from the default
* Trx changed to Tx
* Adding info about the problem value in exceptions
* Reducing delegate use in DeleteExpiredDocumentsCommand <=> Dto
* No need checks in dto => command removed
* Unnecessary async use removed
* Explanation about connection close added
* Modify replay report frequency
* Tupple properties change to pascal
* Version check in reply changed

ayende commented on src/Raven.Client/Json/Converters/LazyStringValueJsonConverter.cs:
```diff
@@ -18,7 +18,8 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
             {
-                throw new SerializationException($"{nameof(BlittableJsonReader)} must to be used for convert to {nameof(BlittableJsonReaderObject)}");
+                throw new SerializationException(
+                    $"Try to read {nameof(LazyStringValue)} property/field by {reader.GetType()} witch is unsuitable reader. Should use {nameof(BlittableJsonReader)}");
```
witch

haludi replied:
done

ayende commented on src/Raven.Client/Json/Converters/StreamConverter.cs:
```diff
@@ -16,17 +16,23 @@ public override void WriteJson(JsonWriter writer, object value, JsonSerializer s
             {
-                throw new SerializationException($"{nameof(BlittableJsonWriter)} must to be used for serialize {nameof(Stream)}");
+                throw new SerializationException(
+                    $"Try to write {nameof(LazyStringValue)} property/field by {writer.GetType()} witch is unsuitable reader. Should use {nameof(BlittableJsonWriter)}");
```
witch

haludi replied:
done

ayende commented on src/Raven.Client/Json/Converters/StreamConverter.cs:
```diff
@@ -16,17 +16,23 @@ public override void WriteJson(JsonWriter writer, object value, JsonSerializer s
             if (!(value is Stream stream))
             {
-                throw new SerializationException($"Try to write non {nameof(Stream)} value with {nameof(StreamConverter)}");
+                throw new SerializationException($"Try to write {value.GetType()} witch is not {nameof(Stream)} value with {nameof(StreamConverter)}");
```
witch

haludi replied:
done

ayende commented on src/Raven.Client/Json/Converters/StreamConverter.cs:
```diff
@@ -35,7 +41,8 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
             {
-                throw new SerializationException($"{nameof(BlittableJsonReader)} must to be used for deserialize {nameof(Stream)}");
+                throw new SerializationException(
+                    $"Try to read {nameof(Stream)} property/field by {reader.GetType()} witch is unsuitable reader. Should use {nameof(BlittableJsonReader)}");
```
witch

haludi replied:
done

ayende commented on src/Raven.Server/Documents/Handlers/BatchHandler.cs:
```diff
@@ -606,9 +606,14 @@ private CollectionName GetFirstConflictCollection(DocumentsOperationContext cont
+
+            public override TransactionOperationsMerger.IReplayableCommandDto<TransactionOperationsMerger.MergedTransactionCommand> ToDto(JsonOperationContext context)
+            {
+                throw new NotImplementedException();
```
Why?

haludi replied:
Just for compiling

haludi replied:
For compiling only

haludi replied:
done

ayende commented on src/Raven.Server/Documents/Handlers/CountersHandler.cs:
```diff
@@ -199,6 +199,11 @@ void LoadDocument()
 
+            public override TransactionOperationsMerger.IReplayableCommandDto<TransactionOperationsMerger.MergedTransactionCommand> ToDto(JsonOperationContext context)
+            {
+                throw new NotImplementedException();
```
Why?

haludi replied:
Just for compiling

ayende replied:
We need to actually implement these, no?

haludi replied:
For compiling only

haludi replied:
done

ayende replied:
Okay, so that was the first pass, then you are going to impl these?

haludi replied:
done

haludi committed:
RavenDB-11373 The ownership of the record stream moved to TxMerger for handle creating file error

haludi committed:
RavenDB-11373 Replay Tx logic moved to helper

haludi committed:
RavenDB-11373 Added an option to record and run ExecuteCounterBatchCommand

haludi committed:
RavenDB-11373 Test for counter operations as batch
and correction of a spelling error on converters errors

maximburyak commented on src/Raven.Client/Json/Converters/BlittableJsonConverter.cs:
```diff
@@ -0,0 +1,55 @@
+        {
+            if (!(reader is BlittableJsonReader blittableReader))
+            {
+                throw new SerializationException(
```
It's better to move the throws to seperate functions

haludi replied:
I move the check to a base class and I think now it's better.
What do you think?

maximburyak commented on src/Raven.Client/Json/Converters/StreamConverter.cs:
```diff
@@ -0,0 +1,66 @@
+            stream.Seek(0, SeekOrigin.Begin);
+            var buffer = stream.ReadData();
+            stream.Seek(0, SeekOrigin.Begin);
+            var strBuffer = Convert.ToBase64String(buffer);
```
Maybe we could consider reducing allocations here, by using LazyStringValue, that is being fed by Convert.ToBase64CharArray

haludi replied:
An issue created
[RavenDB-11852](http://issues.hibernatingrhinos.com/issue/RavenDB-11852)

maximburyak commented on src/Raven.Server/Documents/Expiration/ExpiredDocumentsCleaner.cs:
```diff
@@ -123,10 +124,37 @@ public DeleteExpiredDocumentsCommand(Dictionary<Slice, List<(Slice LowerId, Lazy
+
+            public override TransactionOperationsMerger.IReplayableCommandDto<TransactionOperationsMerger.MergedTransactionCommand> ToDto(JsonOperationContext context)
+            {
+                var keyValuePairs = new List<KeyValuePair<Slice, List<(Slice LowerId, LazyStringValue Id)>>>();
```
We could use an Array here and call Array.Copy, rather than using list

haludi replied:
done

haludi committed:
RavenDB-11373 Added
an option to serialize & deserialize BlittableJsonReaderArray
and an option to record & run ExecuteCounterBatchCommand

haludi committed:
RavenDB-11373 JsonConverters finishings, clone option added to BlittableJsonArrayReader and used in serialization and list use replace by an array in DeleteExpiredDocumentsCommandDto

haludi committed:
Merge branch 'v4.1' of https://github.com/ravendb/ravendb into FinalRecordingMergedTransactionCommand

haludi committed:
RavenDB-11373 Unnecessary comment removed from BlittableJsonReaderObject.CloneOnTheSameContext
